Allow case-insensitive match of :term: references through intersphinx

tomoinn opened the pull request:
Subject: Allow case-insensitive match of :term: references through intersphinx

### Feature or Bugfix

- Bugfix

### Purpose
- #9291 describes a problem where term references are resolved in a case insensitive fashion when matching a locally defined glossary, but when that same glossary is accessed through intersphinx this stops working and the term can only be matched with exact case. I'm not particularly familiar with the architecture of sphinx so I might well have put this in the wrong place, but it does fix this issue.

### Detail

Simple code change, detects that the objtype is `std:term`, builds a dict of lower case to original case for all keys in the inventory under inspection and uses that to perform case insensitive matching.

### Relates
- #9291

tomoinn committed:
Fix for sphinx-doc/sphinx#9291

AWhetter commented on sphinx/ext/intersphinx.py:
```diff
@@ -305,9 +305,19 @@ def missing_reference(app: Sphinx, env: BuildEnvironment, node: pending_xref,
-            proj, version, uri, dispname = inventory[objtype][target]
+            # Special case handling for term to search in a case insensitive fashion
+            if objtype == 'std:term' and objtype in inventory:
+                cased_keys = {k.lower(): k for k in inventory[objtype].keys()}
```
We could avoid creating this whole dictionary by checking each key in turn?

```
lower_target = target.lower()
if any(lower_target == inv_key.lower() for inv_key in inventory[objtype]):
    ...
```

We could avoid doing even this if a case sensitive search returns a result?

```python
if objtype == 'std:term' and objtype in inventory:
    if target in inventory[objtype]:
        proj, version, uri, dispname = inventory[objtype][target]
    else:
        lower_target = target.lower()
        if any(lower_target == inv_key.lower() for inv_key in inventory[objtype]):
            ...
       else:
            continue
else:
    ...
```

tomoinn replied:
I think you still have to build the dict, because after determining that there *is* a case insensitive match you have to look up the target, which is inherently a case sensitive process as it's just a dict key? That or something like

```python
            if objtype not in inventory:
                continue
            if target in inventory[objtype]:
                # Case sensitive match
                proj, version, uri, dispname = inventory[objtype][target]
            elif objtype == 'std:term':
                # Check for potential case insensitive match
                target_lower = target.lower()
                insensitive_matches = list(filter(lambda k: k.lower() == target_lower, inventory[objtype].keys()))
                if insensitive_matches:
                    proj, version, uri, dispname = inventory[objtype][insensitive_matches[0]]
                else:
                    continue
            else:
                continue
```

but I'm not sure this is any more efficient. I guess it's building a list rather than a dict.

AWhetter replied:
Oh you're totally right :facepalm: This list approach will save a small amount of memory at least, plus it gives us the opportunity to raise a warning if multiple keys match.
I like the structure of the `if` statements that you're proposing too!

tomoinn replied:
I've pushed a commit with the above (plus comments). I guess the argument for doing this with a dict is if there are a lot of cases and you can move the dict generation out of the per-lookup loop, so have it outside of the `for inventory, target in to_try:` block, but you still have the memory occupancy to deal with. Priorities, as always!

tomoinn committed:
Somewhat more efficient fix.

Only looks for case insensitive match if there isn't a case sensitive one, and uses filter to build a list of case insensitive matches rather than building a dict.

tk0miya commented:
LGTM! Could you add a testcase for this, please?

tomoinn commented:
I haven't abandoned this - been busy with the project that needs the fix, will add test cases when I have some time available.

tk0miya commented:
Now I added a testcase for this and merged. Thank you for your contribution!

GonzaloT23 reviewed (commented):
Ok
